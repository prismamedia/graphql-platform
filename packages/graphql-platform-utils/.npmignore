.pnp.*
.vscode/extensions.json
.yarn/

**/*.log

node_modules/
tmp/
tsconfig.tsbuildinfo