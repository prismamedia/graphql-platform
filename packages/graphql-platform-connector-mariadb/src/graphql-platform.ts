import {
  BaseContext as CoreBaseContext,
  Context as CoreContext,
  CustomContext,
  CustomOperationConfig as CoreCustomOperationConfig,
  GraphQLPlatform as CoreGraphQLPlatform,
  GraphQLPlatformConfig as CoreGraphQLPlatformConfig,
} from '@prismamedia/graphql-platform-core';
import { Maybe, POJO } from '@prismamedia/graphql-platform-utils';
import { Memoize } from '@prismamedia/ts-memoize';
import {
  Connector,
  ConnectorConfig,
  ConnectorRequest,
} from './graphql-platform/connector';
import { ResourceConfig } from './graphql-platform/resource';

export * from './graphql-platform/connector';
export * from './graphql-platform/resource';

/** "Context" provided by the GraphQL Platform */
export type BaseContext = CoreBaseContext & {
  /**
   * Use the "Connector" at your own risk as you can do whatever you want with the database
   */
  connector: Connector;

  /**
   * A new "ConnectorRequest" is created for every GraphQL execution's context, it is a convenient place
   * to share a state/cache between all the resolvers of the same request
   */
  connectorRequest: ConnectorRequest;
};

export type AnyBaseContext = BaseContext;

export type Context<TCustomContext extends CustomContext = {}> = CoreContext<
  TCustomContext,
  BaseContext
>;

export type CustomOperationConfig<
  TArgs extends POJO = any,
  TCustomContext extends CustomContext = {}
> = CoreCustomOperationConfig<TArgs, TCustomContext, BaseContext>;

export interface GraphQLPlatformConfig<
  TContextParams extends POJO = any,
  TCustomContext extends CustomContext = {},
  TBaseContext extends AnyBaseContext = BaseContext
> extends CoreGraphQLPlatformConfig<
    TContextParams,
    TCustomContext,
    TBaseContext,
    ResourceConfig<TCustomContext, TBaseContext>
  > {
  connector?: Maybe<ConnectorConfig>;
}

export class GraphQLPlatform<
  TContextParams extends POJO = any,
  TCustomContext extends CustomContext = {},
  TBaseContext extends AnyBaseContext = BaseContext
> extends CoreGraphQLPlatform<
  TContextParams,
  TCustomContext,
  TBaseContext,
  GraphQLPlatformConfig<TContextParams, TCustomContext, TBaseContext>
> {
  @Memoize()
  public getConnector(): Connector {
    return new Connector(this.config.connector, this);
  }

  public async getBaseContext(): Promise<TBaseContext> {
    return {
      ...(await super.getBaseContext()),
      connector: this.getConnector(),
      connectorRequest: this.getConnector().newRequest(),
    };
  }
}
