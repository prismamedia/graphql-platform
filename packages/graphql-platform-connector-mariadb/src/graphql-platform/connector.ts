import {
  AnyGraphQLPlatform,
  ConnectorCountOperationArgs,
  ConnectorCreateOperationArgs,
  ConnectorDeleteOperationArgs,
  ConnectorFindOperationArgs,
  ConnectorInterface,
  ConnectorOperationParams as CoreConnectorOperationParams,
  ConnectorUpdateOperationArgs,
  CustomContext,
  OperationEventKind,
} from '@prismamedia/graphql-platform-core';
import { GraphQLOperationType, Maybe, MaybeArray, POJO } from '@prismamedia/graphql-platform-utils';
import EventEmitter from '@prismamedia/ts-async-event-emitter';
import * as mysql from 'mysql';
import { Memoize } from 'typescript-memoize';
import { promisify } from 'util';
import { BaseContext } from '../graphql-platform';
import { Database } from './connector/database';
import { OperationEvent } from './connector/database/operation';
import { ConnectorRequest } from './connector/request';

export * from './connector/database';
export * from './connector/request';

export enum ConnectorEventKind {
  PreStartTransaction = 'PRE_START_TRANSACTION',
  PreRollbackTransaction = 'PRE_ROLLBACK_TRANSACTION',
  PreCommitTransaction = 'PRE_COMMIT_TRANSACTION',
  PreQuery = 'PRE_QUERY',
}

export type ConnectorEventMap = {
  [ConnectorEventKind.PreStartTransaction]: Readonly<Pick<mysql.Connection, 'threadId'>>;
  [ConnectorEventKind.PreRollbackTransaction]: Readonly<Pick<mysql.Connection, 'threadId'>>;
  [ConnectorEventKind.PreCommitTransaction]: Readonly<Pick<mysql.Connection, 'threadId'>>;
  [ConnectorEventKind.PreQuery]: Readonly<Pick<mysql.Connection, 'threadId'> & mysql.QueryOptions>;
};

export type ConnectorConfig = Omit<
  mysql.PoolConfig,
  | 'bigNumberStrings'
  | 'charset'
  | 'dateStrings'
  | 'debug'
  | 'flags'
  | 'insecureAuth'
  | 'multipleStatements'
  | 'queryFormat'
  | 'stringifyObjects'
  | 'supportBigNumbers'
  | 'timezone'
  | 'trace'
  | 'typeCast'
> & {
  /**
   * If you need to set session variables on the connection before it gets used, you're able to do it like:
   * onConnect: connection => connection.query('SET wait_timeout=1; SET max_statement_time=1;')
   */
  onConnect?: Maybe<(connection: mysql.PoolConnection) => void>;
};

export type QueryResult =
  | POJO[]
  | {
      affectedRows: Maybe<number>;
      changedRows: Maybe<number>;
      insertId: Maybe<number>;
    };

export type ConnectorOperationParams<
  TArgs extends POJO,
  TCustomContext extends CustomContext = {}
> = CoreConnectorOperationParams<TArgs, TCustomContext, BaseContext>;

export class Connector<TCustomContext extends CustomContext = {}> extends EventEmitter<ConnectorEventMap>
  implements ConnectorInterface<TCustomContext, BaseContext> {
  protected pool?: mysql.Pool;

  public constructor(readonly config: Maybe<ConnectorConfig>, readonly gp: AnyGraphQLPlatform) {
    super();

    // Handle connection & transaction
    this.gp.getResourceSet().forEach(resource =>
      resource.onConfig({
        [OperationEventKind.PreOperation]: async (event: OperationEvent<any, TCustomContext>) => {
          const {
            operation: { type },
            context: { operationEventDataMap, connectorRequest, logger },
          } = event;

          const eventData = operationEventDataMap.get(event);

          if (type === GraphQLOperationType.Mutation && !connectorRequest.connection) {
            connectorRequest.connection = await this.getConnection();

            try {
              await this.emit(ConnectorEventKind.PreStartTransaction, {
                threadId: connectorRequest.connection.threadId,
              });
            } catch (error) {
              logger && logger.error(logger);
            }

            await promisify(connectorRequest.connection.beginTransaction.bind(connectorRequest.connection))();

            if (eventData) {
              eventData.createdConnection = true;
            }
          }
        },
        [OperationEventKind.PostOperationSuccess]: async (event: OperationEvent<any, TCustomContext>) => {
          const {
            operation: { type },
            context: { operationEventDataMap, connectorRequest, logger },
          } = event;

          const eventData = operationEventDataMap.get(event);

          if (
            type === GraphQLOperationType.Mutation &&
            connectorRequest.connection &&
            eventData &&
            eventData.createdConnection === true
          ) {
            try {
              await this.emit(ConnectorEventKind.PreCommitTransaction, {
                threadId: connectorRequest.connection.threadId,
              });
            } catch (error) {
              logger && logger.error(logger);
            }

            await promisify(connectorRequest.connection.commit.bind(connectorRequest.connection))();
          }
        },
        [OperationEventKind.PostOperationError]: async (event: OperationEvent<any, TCustomContext>) => {
          const {
            operation: { type },
            context: { operationEventDataMap, connectorRequest, logger },
          } = event;

          const eventData = operationEventDataMap.get(event);

          if (
            type === GraphQLOperationType.Mutation &&
            connectorRequest.connection &&
            eventData &&
            eventData.createdConnection === true
          ) {
            try {
              await this.emit(ConnectorEventKind.PreRollbackTransaction, {
                threadId: connectorRequest.connection.threadId,
              });
            } catch (error) {
              logger && logger.error(logger);
            }

            await promisify(connectorRequest.connection.rollback.bind(connectorRequest.connection))();
          }
        },
        [OperationEventKind.PostOperation]: async (event: OperationEvent<any, TCustomContext>) => {
          const {
            operation: { type },
            context: { operationEventDataMap, connectorRequest },
          } = event;

          const eventData = operationEventDataMap.get(event);

          if (
            type === GraphQLOperationType.Mutation &&
            connectorRequest.connection &&
            eventData &&
            eventData.createdConnection === true
          ) {
            connectorRequest.connection.release();

            delete connectorRequest.connection;
          }
        },
      }),
    );
  }

  @Memoize()
  public getCharset(): string {
    return 'utf8mb4';
  }

  @Memoize()
  public getCollation(): string {
    return 'utf8mb4_unicode_520_ci';
  }

  @Memoize()
  public getDatabase(): Database {
    return new Database(this);
  }

  protected getPool(): mysql.Pool {
    if (this.pool) {
      return this.pool;
    }

    if (!this.config) {
      throw new Error(`You have to provide the connector's configuration.`);
    }

    const { onConnect, ...poolConfig } = this.config;

    return (this.pool = mysql
      .createPool({
        ...poolConfig,
        charset: this.getCollation(),
        debug: false,
        multipleStatements: false,
        supportBigNumbers: true,
        timezone: 'Z',
        waitForConnections: true,
        dateStrings: false,
      })
      .on('connection', connection => {
        // We ensure the good charset and the good timezone to be used
        connection.query(`SET NAMES ${this.getCharset()} COLLATE ${this.getCollation()}; SET time_zone = 'UTC';`);

        // We let the user configure the connection
        onConnect && onConnect(connection);
      }));
  }

  public async closePool(): Promise<void> {
    if (this.pool) {
      await promisify(this.pool.end.bind(this.pool))();
      delete this.pool;
    }
  }

  public async resetPool(): Promise<void> {
    await this.closePool();
    this.getPool();
  }

  protected async getConnection(): Promise<mysql.PoolConnection> {
    const pool = this.getPool();

    return promisify(pool.getConnection.bind(pool))();
  }

  /** Ensures the connection is released at the end of the task. */
  public async withConnection<R>(task: (connection: mysql.Connection) => Promise<R>): Promise<R> {
    let result: R;

    const connection = await this.getConnection();
    try {
      result = await task(connection);
    } finally {
      connection.release();
    }

    return result;
  }

  public async query(
    query: MaybeArray<mysql.QueryOptions | mysql.QueryOptions['sql']>,
    maybeConnection?: Maybe<mysql.Connection | BaseContext>,
  ): Promise<QueryResult> {
    const connection: mysql.Connection | undefined = maybeConnection
      ? 'connectorRequest' in maybeConnection
        ? maybeConnection.connectorRequest.connection
        : maybeConnection
      : undefined;

    return connection
      ? Array.isArray(query)
        ? // As we use only one connection, even if they are thrown concurrently, they are executed serially
          Promise.all(query.map(async query => this.query(query, connection)))
        : new Promise(async (resolve, reject) => {
            const parsedQuery: mysql.QueryOptions = typeof query === 'string' ? { sql: query } : query;

            await this.emit(ConnectorEventKind.PreQuery, {
              threadId: connection.threadId,
              ...parsedQuery,
            });

            connection.query(query, async (error, rows) => (error ? reject(error) : resolve(rows)));
          })
      : this.withConnection(async connection => this.query(query, connection));
  }

  public newRequest(): ConnectorRequest {
    return new ConnectorRequest(this);
  }

  public async find({ resource, ...params }: ConnectorOperationParams<ConnectorFindOperationArgs>) {
    return this.getDatabase()
      .getTable(resource)
      .getOperation('Find')
      .execute(Object.freeze(params));
  }

  public async count({ resource, ...params }: ConnectorOperationParams<ConnectorCountOperationArgs>) {
    return this.getDatabase()
      .getTable(resource)
      .getOperation('Count')
      .execute(Object.freeze(params));
  }

  public async create({ resource, ...params }: ConnectorOperationParams<ConnectorCreateOperationArgs>) {
    return this.getDatabase()
      .getTable(resource)
      .getOperation('Create')
      .execute(Object.freeze(params));
  }

  public async update({ resource, ...params }: ConnectorOperationParams<ConnectorUpdateOperationArgs>) {
    return this.getDatabase()
      .getTable(resource)
      .getOperation('Update')
      .execute(Object.freeze(params));
  }

  public async delete({ resource, ...params }: ConnectorOperationParams<ConnectorDeleteOperationArgs>) {
    return this.getDatabase()
      .getTable(resource)
      .getOperation('Delete')
      .execute(Object.freeze(params));
  }
}
