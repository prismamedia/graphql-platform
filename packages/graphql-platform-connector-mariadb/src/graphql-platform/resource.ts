import {
  CustomContext,
  Resource as CoreResource,
  ResourceConfig as CoreResourceConfig,
} from '@prismamedia/graphql-platform-core';
import { Maybe } from '@prismamedia/graphql-platform-utils';
import { BaseContext } from '../graphql-platform';
import { TableConfig } from './connector/database/table';
import { FieldConfig, RelationConfig } from './resource/component';
import { UniqueFullConfig } from './resource/unique';

export * from './resource/component';
export * from './resource/unique';

export interface ResourceConfig<TCustomContext extends CustomContext = any>
  extends CoreResourceConfig<
    TCustomContext,
    BaseContext,
    FieldConfig<TCustomContext>,
    RelationConfig<TCustomContext>,
    UniqueFullConfig
  > {
  table?: Maybe<TableConfig>;
}

export type Resource = CoreResource<ResourceConfig>;
