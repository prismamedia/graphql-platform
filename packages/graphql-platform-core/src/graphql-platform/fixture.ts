import { Entries, fromEntries } from '@prismamedia/graphql-platform-utils';
import { DepGraph } from 'dependency-graph';
import { OperationDefinitionNode, print } from 'graphql';
import { Logger } from 'winston';
import { GraphQLPlatform, GraphQLRequest } from '../graphql-platform';
import { CreateOneDataInputValue } from './operation';
import { AnyResource, Relation, SerializedFieldValue } from './resource';
import { TypeKind, WhereUniqueInputValue } from './type';

export class FixtureGraph extends DepGraph<Fixture> {}

export type FixtureFieldData = SerializedFieldValue;

export type FixtureRelationData = null | Fixture['name'];

export type FixtureComponentData = FixtureFieldData | FixtureRelationData;

export type FixtureData = {
  [componentName: string]: FixtureComponentData;
};

export type FixtureDataMap = {
  [fixtureName: string]: FixtureData;
};

export class Fixture {
  readonly logger?: Logger;
  protected id?: WhereUniqueInputValue;

  public constructor(
    readonly name: string,
    readonly resource: AnyResource,
    readonly data: FixtureData,
    protected graph: FixtureGraph,
    protected gp: GraphQLPlatform,
  ) {
    this.logger = gp.getLogger();
  }

  public toString(): string {
    return `${this.resource}.${this.name}`;
  }

  public getRelatedFixture(relation: Relation): Fixture | null {
    this.resource.getRelationMap().assert(relation);

    const relationValue = this.data[relation.name];

    if (relationValue == null) {
      if (relation.isRequired()) {
        throw new Error(`In the fixture "${this.name}", the relation "${relation}" can't be null.`);
      }
    } else if (!(typeof relationValue === 'string' && this.graph.hasNode(relationValue))) {
      throw new Error(
        `In the fixture "${this.name}", the relation "${relation}" has to reference an "${relation.getTo()}" fixture.`,
      );
    }

    return relationValue ? this.graph.getNodeData(relationValue) : null;
  }

  public getCreateMutationAST(): OperationDefinitionNode {
    const operationDefinitionNode: OperationDefinitionNode = {
      kind: 'OperationDefinition',
      operation: 'mutation',
      variableDefinitions: [
        {
          kind: 'VariableDefinition',
          variable: {
            kind: 'Variable',
            name: {
              kind: 'Name',
              value: 'data',
            },
          },
          type: {
            kind: 'NonNullType',
            type: {
              kind: 'NamedType',
              name: {
                kind: 'Name',
                value: this.resource.getMutation('CreateOne').getDataType().name,
              },
            },
          },
        },
      ],
      selectionSet: {
        kind: 'SelectionSet',
        selections: [
          {
            kind: 'Field',
            name: {
              kind: 'Name',
              value: this.resource.getMutation('CreateOne').name,
            },
            alias: {
              kind: 'Name',
              value: 'id',
            },
            arguments: [
              {
                kind: 'Argument',
                name: {
                  kind: 'Name',
                  value: 'data',
                },
                value: {
                  kind: 'Variable',
                  name: {
                    kind: 'Name',
                    value: 'data',
                  },
                },
              },
            ],
            selectionSet: this.resource
              .getFirstPublicUnique()
              .getSelectionNode(TypeKind.Output)
              .toAST(),
          },
        ],
      },
    };

    return operationDefinitionNode;
  }

  public getCreateMutationSource(): string {
    return print(this.getCreateMutationAST());
  }

  public getCreateMutationVariables(): GraphQLRequest['variableValues'] {
    return {
      data: fromEntries([
        // Fields
        ...[...this.resource.getFieldSet()].map(field =>
          !field.isFullyManaged() ? [field.name, this.data[field.name] as FixtureFieldData] : undefined,
        ),
        // Relations
        ...[...this.resource.getRelationSet()].map(relation => {
          if (!relation.isFullyManaged()) {
            const relatedFixture = this.getRelatedFixture(relation);
            if (relatedFixture) {
              return [relation.name, { connect: relatedFixture.assertId() }];
            }
          }
        }),
      ] as Entries<CreateOneDataInputValue>),
    };
  }

  public getCreateMutation(): GraphQLRequest {
    return {
      source: this.getCreateMutationSource(),
      variableValues: this.getCreateMutationVariables(),
    };
  }

  public async load(contextValue?: GraphQLRequest['contextValue']): Promise<void> {
    try {
      const { id } = await this.gp.execute<{ id: WhereUniqueInputValue }>({
        ...this.getCreateMutation(),
        contextValue,
      });

      this.id = this.resource.getInputType('WhereUnique').assert(id);
    } catch (error) {
      this.logger && this.logger.error(error);

      throw error;
    }
  }

  public assertId(): WhereUniqueInputValue {
    if (!this.id) {
      throw new Error(`The fixture "${this.name}" is not loaded yet.`);
    }

    return this.id;
  }
}
