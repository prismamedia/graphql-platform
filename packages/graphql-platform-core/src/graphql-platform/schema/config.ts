import { GraphQLOperationType } from '@prismamedia/graphql-platform-utils';
import { GraphQLFieldConfigMap, GraphQLObjectType, GraphQLSchemaConfig as BaseGraphQLSchemaConfig } from 'graphql';
import { OperationType } from '../operation';
import { ResourceMap } from '../resource/map';

export interface GraphQLSchemaConfigConfig
  extends Omit<BaseGraphQLSchemaConfig, 'mutation' | 'query' | 'subscription'> {
  resourceMap: ResourceMap;
  queryMap: GraphQLFieldConfigMap<any, any>;
  mutationMap: GraphQLFieldConfigMap<any, any>;
}

export class GraphQLSchemaConfig implements BaseGraphQLSchemaConfig {
  public constructor(readonly config: GraphQLSchemaConfigConfig) {}

  public getOperationTypeFieldMap<TType extends OperationType>(type: TType): GraphQLFieldConfigMap<any, any> {
    const fields: GraphQLFieldConfigMap<any, any> = {};

    for (const resource of this.config.resourceMap.values()) {
      for (const operation of resource.getOperations(type)) {
        if (operation.isPublic()) {
          fields[operation.name] = operation.getGraphQLFieldConfig();
        }
      }
    }

    return fields;
  }

  public get mutation() {
    return new GraphQLObjectType({
      name: 'Mutation',
      fields: { ...this.getOperationTypeFieldMap(GraphQLOperationType.Mutation), ...this.config.mutationMap },
    });
  }

  public get query() {
    return new GraphQLObjectType({
      name: 'Query',
      fields: { ...this.getOperationTypeFieldMap(GraphQLOperationType.Query), ...this.config.queryMap },
    });
  }

  public get subscription() {
    return undefined;
  }
}
