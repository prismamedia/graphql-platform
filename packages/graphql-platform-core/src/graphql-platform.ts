import {
  FlagConfig,
  getFlagValue,
  getThunkValue,
  loadModuleMap,
  Maybe,
  MaybePromise,
  Merge,
  mergeWith,
  ModuleMapConfig,
  POJO,
  Thunk,
} from '@prismamedia/graphql-platform-utils';
import { graphql, GraphQLArgs, GraphQLFieldConfig, GraphQLNamedType, GraphQLSchema } from 'graphql';
import { Binding } from 'graphql-binding';
import { ExecutionResultDataDefault } from 'graphql/execution/execute';
import { Memoize } from 'typescript-memoize';
import { Logger } from 'winston';
import { ConnectorInterface } from './graphql-platform/connector';
import { Fixture, FixtureData, FixtureGraph } from './graphql-platform/fixture';
import { OperationContext, OperationEvent } from './graphql-platform/operation';
import {
  AnyResourceConfig,
  MaybeResourceMapAware,
  Resource,
  ResourceConfig,
  ResourceGraph,
  ResourceMap,
  ResourceSet,
} from './graphql-platform/resource';
import { GraphQLSchemaConfig } from './graphql-platform/schema';

export * from './graphql-platform/connector';
export * from './graphql-platform/fixture';
export * from './graphql-platform/operation';
export * from './graphql-platform/resource';
export * from './graphql-platform/schema';
export * from './graphql-platform/type';

/** "Context" provided by the GraphQL Platform */
export type BaseContext = {
  /**
   * The logger, if any
   */
  logger: Logger | undefined;

  /**
   * Return the current environment
   */
  environment: string;

  /**
   * Either we are in debug mode or not
   */
  debug: boolean;

  /**
   * This GraphQL Platform's resources
   */
  resourceMap: ResourceMap;

  /**
   * This API's binding, in order to execute other requests
   */
  api: Binding;

  /**
   * This operation's context
   */
  operationContext: OperationContext;

  /**
   * A convenient location to store data between the different operations' events
   */
  operationEventDataMap: WeakMap<OperationEvent, any>;
};

export type AnyBaseContext = BaseContext;

/** "Context" provided by the GraphQL Platform's user, can't override the "BaseContext"'s properties */
export interface CustomContext {
  [key: string]: any;
}

/** "Context" available in the operations, hooks... */
export type Context<
  TCustomContext extends CustomContext = {},
  TBaseContext extends AnyBaseContext = BaseContext
> = TCustomContext & TBaseContext;

export type GraphQLRequest = Merge<Omit<GraphQLArgs, 'schema'>, { contextValue?: CustomContext }>;

export interface GraphQLExecutor {
  execute<TData = ExecutionResultDataDefault>(request: GraphQLRequest): Promise<TData>;
}

export type CustomOperationConfig<
  TArgs extends POJO = any,
  TCustomContext extends CustomContext = {},
  TBaseContext extends AnyBaseContext = BaseContext
> = MaybeResourceMapAware<GraphQLFieldConfig<any, Context<TCustomContext, TBaseContext>, TArgs>>;

export interface GraphQLPlatformConfig<
  TContextParams extends POJO = any,
  TCustomContext extends CustomContext = {},
  TBaseContext extends AnyBaseContext = BaseContext,
  TResourceConfig extends AnyResourceConfig = ResourceConfig<TCustomContext, TBaseContext>
> {
  /** Optional, provide your own logger to see what happens under the hood */
  logger?: Maybe<Logger>;

  /** Optional, default: true but in "production" and "test" env */
  debug?: FlagConfig;

  /** Optional, provide a context you can use in the virtual fields and custom operations' resolver */
  context?: Maybe<(params?: TContextParams) => MaybePromise<TCustomContext>>;

  /** Optional, fine-tune all the resources at once */
  default?: Maybe<(resourceName: Resource['name']) => TResourceConfig>;

  /** Optional, provide some resource configs */
  resources?: ModuleMapConfig<TResourceConfig>;

  /** Optional, provide some custom queries */
  queries?: ModuleMapConfig<CustomOperationConfig<any, TCustomContext, TBaseContext>>;

  /** Optional, provide some custom mutations */
  mutations?: ModuleMapConfig<CustomOperationConfig<any, TCustomContext, TBaseContext>>;

  /** Optional, provide some custom types */
  types?: Maybe<MaybeResourceMapAware<GraphQLNamedType[]>>;

  /** Optional, default fixtures */
  fixtures?: Thunk<ModuleMapConfig<ModuleMapConfig<FixtureData>>>;
}

export type AnyGraphQLPlatformConfig = GraphQLPlatformConfig<any, any, any, any>;

export class GraphQLPlatform<
  TContextParams extends POJO = any,
  TCustomContext extends CustomContext = {},
  TConfig extends AnyGraphQLPlatformConfig = GraphQLPlatformConfig<TContextParams, TCustomContext>
> implements GraphQLExecutor {
  public constructor(readonly config: TConfig) {}

  @Memoize()
  public getLogger(): Logger | undefined {
    return this.config.logger || undefined;
  }

  @Memoize()
  public getEnvironment(): string {
    return process.env.NODE_ENV || 'development';
  }

  @Memoize()
  public isDebug(): boolean {
    return getFlagValue(this.config.debug, !['production', 'test'].includes(this.getEnvironment()));
  }

  @Memoize()
  public getResourceMap(): ResourceMap {
    const resourceMap = new ResourceMap();

    for (const [resourceName, resourceConfig] of loadModuleMap(this.config.resources)) {
      // Generate the default config by providing the resource's name
      const defaultResourceConfig = this.config.default ? this.config.default(resourceName) : undefined;

      // Process the configs merging, the resource's config prevails over the default one.
      const mergedResourceConfig = defaultResourceConfig
        ? mergeWith({}, defaultResourceConfig, resourceConfig)
        : resourceConfig;

      resourceMap.set(resourceName, new Resource(resourceName, mergedResourceConfig, this));
    }

    return resourceMap;
  }

  @Memoize()
  public getResourceSet(): ResourceSet {
    return new ResourceSet(this.getResourceMap().values());
  }

  @Memoize()
  public getResourceGraph(): ResourceGraph {
    const graph = new ResourceGraph({ circular: false });

    for (const resource of this.getResourceSet()) {
      graph.addNode(resource.name, resource);
    }

    for (const resource of this.getResourceSet()) {
      for (const relation of resource.getRelationSet()) {
        if (relation.getFrom() !== relation.getTo()) {
          graph.addDependency(relation.getFrom().name, relation.getTo().name);
        }
      }
    }

    return graph;
  }

  @Memoize()
  public getGraphQLSchemaConfig(): GraphQLSchemaConfig {
    const resourceMap = this.getResourceMap();

    return new GraphQLSchemaConfig({
      resourceMap,
      queryMap: this.config.queries
        ? [...loadModuleMap(this.config.queries)].reduce(
            (operationMap, [name, config]) =>
              Object.assign(operationMap, { [name]: typeof config === 'function' ? config({ resourceMap }) : config }),
            {},
          )
        : {},
      mutationMap: this.config.mutations
        ? [...loadModuleMap(this.config.mutations)].reduce(
            (operationMap, [name, config]) =>
              Object.assign(operationMap, { [name]: typeof config === 'function' ? config({ resourceMap }) : config }),
            {},
          )
        : {},
      types: this.config.types
        ? typeof this.config.types === 'function'
          ? this.config.types({ resourceMap })
          : this.config.types
        : [],
    });
  }

  @Memoize()
  public getGraphQLSchema(): GraphQLSchema {
    return new GraphQLSchema(this.getGraphQLSchemaConfig());
  }

  @Memoize()
  public getBinding(): Binding {
    return new Binding({ schema: this.getGraphQLSchema() });
  }

  public async getCustomContext(params?: any): Promise<CustomContext> {
    return this.config.context ? this.config.context(params) : {};
  }

  public async getBaseContext(): Promise<BaseContext> {
    return {
      logger: this.getLogger(),
      environment: this.getEnvironment(),
      debug: this.isDebug(),
      resourceMap: this.getResourceMap(),
      api: this.getBinding(),
      operationContext: {},
      operationEventDataMap: new WeakMap(),
    };
  }

  public async getContext(params?: any): Promise<Context<CustomContext, BaseContext>> {
    const [custom, base] = await Promise.all([this.getCustomContext(params), this.getBaseContext()]);

    return Object.freeze({
      ...custom,
      ...base,
    });
  }

  public async execute<TData = ExecutionResultDataDefault>(request: GraphQLRequest): Promise<TData> {
    const { data, errors } = await graphql<TData>({
      ...request,
      schema: this.getGraphQLSchema(),
      contextValue: Object.freeze({
        ...(await this.getContext()),
        ...request.contextValue,
      }),
    });

    if (errors && errors.length > 0) {
      throw errors[0];
    }

    if (data == null) {
      throw new Error(
        `An empty result have been returned for the following request: ${JSON.stringify({
          source: request.source,
          variableValues: request.variableValues,
        })}`,
      );
    }

    return data;
  }

  @Memoize()
  public getConnector(): ConnectorInterface {
    return {
      find() {
        throw new Error(`The "find" operation is not implemented, yet.`);
      },
      count() {
        throw new Error(`The "count" operation is not implemented, yet.`);
      },
      create() {
        throw new Error(`The "create" operation is not implemented, yet.`);
      },
      update() {
        throw new Error(`The "update" operation is not implemented, yet.`);
      },
      delete() {
        throw new Error(`The "delete" operation is not implemented, yet.`);
      },
    };
  }

  // Gets the fixtures, sorted by their dependencies
  public getFixtureGraph(
    config: ModuleMapConfig<ModuleMapConfig<FixtureData>> = getThunkValue(this.config.fixtures),
  ): FixtureGraph {
    const fixtureGraph = new FixtureGraph({ circular: false });

    for (const [resourceName, fixtureDataMap] of loadModuleMap(config)) {
      const resource = this.getResourceMap().assert(resourceName);
      for (const [fixtureName, fixtureData] of loadModuleMap(fixtureDataMap)) {
        if (fixtureGraph.hasNode(fixtureName)) {
          const fixture = fixtureGraph.getNodeData(fixtureName);

          throw new Error(`At least 2 fixtures have the same name: "${resource}.${fixtureName}" and "${fixture}"`);
        }

        const fixture = new Fixture(fixtureName, resource, fixtureData, fixtureGraph, this);

        fixtureGraph.addNode(fixture.name, fixture);
      }
    }

    // Sort the fixtures
    for (const fixtureName of fixtureGraph.overallOrder()) {
      const fixture = fixtureGraph.getNodeData(fixtureName);

      for (const relation of fixture.resource.getRelationSet()) {
        const relatedFixture = fixture.getRelatedFixture(relation);
        if (relatedFixture) {
          fixtureGraph.addDependency(fixture.name, relatedFixture.name);
        }
      }
    }

    return fixtureGraph;
  }

  public async loadFixtures(
    config?: ModuleMapConfig<ModuleMapConfig<FixtureData>>,
    contextValue?: GraphQLRequest['contextValue'],
  ): Promise<void> {
    const fixtureGraph = this.getFixtureGraph(config);

    // Actually load the fixtures
    for (const fixtureName of fixtureGraph.overallOrder()) {
      await fixtureGraph.getNodeData(fixtureName).load(contextValue);
    }
  }
}

export type AnyGraphQLPlatform = GraphQLPlatform<any, any, any>;
