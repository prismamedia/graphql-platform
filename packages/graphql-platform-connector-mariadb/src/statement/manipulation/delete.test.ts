import { afterAll, beforeAll, describe, expect, it } from '@jest/globals';
import { Node, NodeChange } from '@prismamedia/graphql-platform';
import { MutationType } from '@prismamedia/graphql-platform-utils';
import {
  ArticleStatus,
  myAdminContext,
} from '@prismamedia/graphql-platform/__tests__/config.js';
import * as fixtures from '@prismamedia/graphql-platform/__tests__/fixture.js';
import { createMyGP, type MyGP } from '../../__tests__/config.js';

describe('Delete statement', () => {
  let gp: MyGP;
  const changes: NodeChange[] = [];

  beforeAll(async () => {
    gp = createMyGP(`connector_mariadb_delete_statement`);
    gp.on('node-change', (change) => changes.push(change));

    await gp.connector.setup();
    await gp.seed(myAdminContext, fixtures.constant);
  });

  afterAll(() => gp.connector.teardown());

  it.each([
    [
      'Article',
      myAdminContext,
      {
        where: { status: ArticleStatus.PUBLISHED },
        orderBy: ['createdAt_ASC'],
        first: 10,
        selection: '{ title }',
      },
    ],
    [
      'Article',
      myAdminContext,
      {
        where: { status: ArticleStatus.DRAFT },
        orderBy: ['createdAt_ASC'],
        first: 10,
        selection: '{ title }',
      },
    ],
  ])('generates statements', async (nodeName, context, args) => {
    changes.length = 0;

    await expect(
      gp
        .getNodeByName(nodeName)
        .getMutationByKey('delete-many')
        .execute(context, args),
    ).resolves.toMatchSnapshot('result');

    expect(
      changes.reduce<Map<Node['name'], Map<MutationType, number>>>(
        (changesByMutationTypeByNodeName, change) => {
          let changesByMutationType = changesByMutationTypeByNodeName.get(
            change.node.name,
          );

          if (!changesByMutationType) {
            changesByMutationTypeByNodeName.set(
              change.node.name,
              (changesByMutationType = new Map()),
            );
          }

          changesByMutationType.set(
            change.kind,
            (changesByMutationType.get(change.kind) ?? 0) + 1,
          );

          return changesByMutationTypeByNodeName;
        },
        new Map(),
      ),
    ).toMatchSnapshot('changes');
  });
});
