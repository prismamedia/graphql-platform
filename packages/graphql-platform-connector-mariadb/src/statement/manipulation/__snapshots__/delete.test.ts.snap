// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Delete statement generates statements 1`] = `
[
  {
    "title": "My first published article",
  },
  {
    "title": "My second published article",
  },
]
`;

exports[`Delete statement generates statements 2`] = `
Map {
  "ArticleTagModeration" => Map {
    "deletion" => 2,
  },
  "ArticleTag" => Map {
    "deletion" => 3,
  },
  "Article" => Map {
    "deletion" => 2,
  },
}
`;

exports[`Delete statement generates statements 3`] = `
[
  {
    "title": "My first draft article",
  },
  {
    "title": "My second draft article",
  },
]
`;

exports[`Delete statement generates statements 4`] = `
Map {
  "Article" => Map {
    "deletion" => 2,
  },
}
`;
